### PR TITLE
Self hosting with docker compose

### PR DESCRIPTION
**Pull request by mike12345567**
## Description
So self hosting has landed....
image

This is the most basic version of self hosting, provided through the use of [docker compose](https://github.com/docker/compose) with a few new services to take over for some of the services we use in AWS.

New services are as follows:
1. [Minio](https://min.io/) - this takes over for S3 locally
2. [Envoy Proxy](https://www.envoyproxy.io/) - this is basically our replacement for Cloudformation, however all requests to the cluster go through here, all requests are handled by the reverse proxy based on URI locations e.g. `localhost:10000/app` would go to the app server.
3. The new worker service - handles validation of deployments and supplying of keys for the deployment process, as we currently do with lambdas, this can be used for more background processes (like automations) and possibly an expanded credentials system (builder users/sharing in self hosting).

I'm not going to document everything here, just give a basic overview to help with review, the plan is to write up full documentation of how to make good use of this.

All of the components that make up the docker compose system can be found in the top level `hosting` directory of the repo. We may want to break these out into their own repo in the future, but for now they need to live nearby as we build the containers from source. Things of note in the hosting directory are as follows:
1. `docker-compose.yaml` file which defines the configuration.
2. `hosting.properties` which contains all the configuration that can be changed about a self hosted cluster.
3. `envoy.yaml` which controls all the reverse proxy configuration.
4. `start.sh` and `install-docker-compose.sh` just some simple shell scripts to simplify the experience on Unix (sorry Windows!)

I have provided the `start.sh` file because we aren't using the default configuration of docker compose, our "up" command has to be `docker-compose --env-file hosting.properties up`.

Also if you make changes to the server or the worker services then you will need to run the command `docker-compose --env-file hosting.properties up --build` to re-build the docker images.

For the most part self hosting is quite simple, the steps are as follows:
1. Install docker compose
2. enter the `hosting` directory
3. Run the up command
4. Point your builder at the self hosted cluster
5. Deploy!

## Further work
1. We can simplify the process of starting up a self hosted cluster by building and pushing docker images to docker hub, meaning that the images don't need to be built from source everytime. This would require us to release the image in the same way as we currently do to ECR as part of our CI/CD.
2. Building up CF config for AWS, and perhaps expanding to cover Kubernetes for more advanced usage. One thing to test is the new AWS capability to use a docker compose file on ECS, this could do the job for the main time.

## Screenshots
Not a lot to screenshot here, apart from a new menu in the builder, the settings menu seen below:
image

And here it is open, configured for self hosting (all inputs shown are the defaults, for a basic usage on the same machine simply toggling to self hosting and running the composed cluster is enough):
image

### CODE REVIEW

**Comment by aptkingston** [comment]
This is class. I'll spin this up on my server and do some real world testing of running this on another host. Might try throwing my own SSL reverse proxy in front of it and having a secure public endpoint too, to see how envoy handles that.

**Review by shogunpurple — APPROVED** [review]
LGTM! Awesome work. Will give this a try in a little cloud server as well to get it up and running. I was thinking it would be nice to create an AMI for this in AWS, and we could use it to quickly bring up some budibase servers in our account for testing.

File: hosting/hosting.properties [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+MAIN_PORT=10000
+
+# Use this password when configuring your self hosting settings
+# This should be updated
+HOSTING_KEY=budibase
+
+# This section contains customisation options
+HOSTING_URL=http://localhost
+LOGO_URL=https://logoipsum.com/logo/logo-15.svg
+HOSTING_URL=http://localhost
```
Comment (aptkingston) [inline hosting/hosting.properties:11]: FATAL ERROR (jk): HOSTING_URL duplicated here it looks like

**Review by aptkingston — APPROVED** [review]
Only outstanding issue I can see is authentication when having multiple deployed apps, but I doubt that's anything particularly to do with self hosting so I don't think it should block merging this in. The issue is basically that if you have a couple of different apps running in different tabs, sometimes logging in/out of one will affect other apps too. Cookies are never nice to use - I would love to replace the auth with JWT's / auth headers / localstorage instead... But again this is way outside of the scope of this particular PR. Self hosting works a treat!

**Pull request merged**